### PR TITLE
Feature/tcache

### PR DESCRIPTION
**Pull request by jtao1735**

### CODE REVIEW

File: src/util/src/tcache.c [hunk 1]
```diff
@@ -19,688 +19,581 @@
...
-  if (next) {
-    next->prev = pElem->prev;
-  }
+  taosWLockLatch(pCacheObj->nlock + hash);
 
-  if (pCacheObj->numOfElemsInTrash == 0) {
-    assert(pCacheObj->pTrash == NULL);
+  SCacheNode *pNode = pCacheObj->nodeList[hash];
+  while (pNode) {
+    if (memcmp(pNode->key, key, keyLen) == 0) break;
```
Comment (localvar) [inline src/util/src/tcache.c:184]: should compare keyLen first

File: src/util/src/tcache.c [hunk 2]
```diff
@@ -19,688 +19,581 @@
...
-  uDebug("cache:%s key:%p, %p move to trashcan, pTrashElem:%p, numOfElem in trashcan:%d", pCacheObj->name,
-      pNode->key, pNode->data, pElem, pCacheObj->numOfElemsInTrash);
+  SCacheObj *pCacheObj = cacheObjList + cacheId;
+  return pCacheObj->count;
 }
 
-void taosTrashcanEmpty(SCacheObj *pCacheObj, bool force) {
-  __cache_wr_lock(pCacheObj);
+static void taosCacheProcessTimer(void *param, void *tmrId) {
+  int64_t cacheId = (int64_t) param;
```
Comment (localvar) [inline src/util/src/tcache.c:463]: cacheId? cacheObj?

File: src/util/src/tcache.c [hunk 3]
```diff
@@ -19,688 +19,581 @@
...
+  SCacheObj *pCacheObj = cacheObjList + cacheId;
+  if (pCacheObj->count * TAOS_CACHE_LOAD_FACTOR > pCacheObj->size && pCacheObj->iterNum <= 0) {
+    taosWLockLatch(&pCacheObj->olock);
+    taosCacheResize(pCacheObj);
+    taosWUnLockLatch(&pCacheObj->olock);
+  }
+ 
+  uint32_t hash = (*pCacheObj->hashFp)(key, (uint32_t)keyLen) % pCacheObj->size;
+  taosRLockLatch(&pCacheObj->olock);
+  if (pCacheObj->state != TAOS_CACHE_STATE_ACTIVE) {
```
Comment (localvar) [inline src/util/src/tcache.c:174]: the check of `pCacheObj->state` is protected by `pCacheObj->olock` here, but by `tsCacheMutex` in `taosCacheCleanup`, which may be a race condition

File: src/util/src/tcache.c [hunk 4]
```diff
@@ -19,688 +19,581 @@
...
-  } else {
-    atomic_add_fetch_32(&pCacheObj->statistics.missCount, 1);
-    uDebug("cache:%s, key:%p, not in cache, retrieved failed", pCacheObj->name, key);
-  }
+  taosWUnLockLatch(pCacheObj->nlock + hash);
+  taosRUnLockLatch(&pCacheObj->olock);
 
-  atomic_add_fetch_32(&pCacheObj->statistics.totalAccess, 1);
-  return pData;
+  if (pCacheObj->state == TAOS_CACHE_STATE_DELETED && count == 0)
```
Comment (localvar) [inline src/util/src/tcache.c:289]: race condition? `pCacheObj->count` could be different with `count`

File: src/util/src/tcache.c [hunk 5]
```diff
@@ -19,688 +19,581 @@
...
 
-  return NULL;
+  pCacheObj->size = newSize;
+
+  int64_t et = taosGetTimestampUs();
+  uDebug("cache:%s, hash table resized, new size:%d, elapsed time:%fms", pCacheObj->name, pCacheObj->size, (et - st)/1000.0); 
 }
+ 
+static void taosCacheResetObj(SCacheObj *pCacheObj) {
+  if (pCacheObj->state == TAOS_CACHE_STATE_EMPTY) return;
```
Comment (localvar) [inline src/util/src/tcache.c:585]: seems there could be concurrent calls to this function, this check won't work in this case.

**Comment by guanshengliang** [comment]
There have been many conflicts, so close this PR

**Pull request closed without merging**